### PR TITLE
Handle MethodInstance keys in interop.unopt

### PR DESCRIPTION
**Pull request by timholy**
Sadly, I don't know enough to easily generate a test case, and it came from Makie which is an undesirably-large test dependency. Any thoughts?

### CODE REVIEW

File: src/Cthulhu.jl [hunk 1]
```diff
@@ -346,10 +347,14 @@ function _descend(term::AbstractTerminal, interp::CthulhuInterpreter, mi::Method
                         (; src, rt, infos, slottypes, codeinf, effects) = lookup(interp, mi, optimize)
                     end
                 else
-                    codeinf = src = copy(interp.unopt[override].src)
-                    rt = interp.unopt[override].rt
-                    infos = interp.unopt[override].stmt_info
-                    effects = get_effects(interp.unopt[override])
+                    unopt = get(interp.unopt, override, missing)
+                    if unopt === missing
+                        unopt = get(interp.unopt, override.linfo, missing)
```
Comment (aviatesk) [inline src/Cthulhu.jl:352]:
```suggestion
                        unopt = interp.unopt[override.linfo]
```
maybe?

**Review by aviatesk — APPROVED** [review]
I'm not too sure when `interp.unopt` doesn't have the given `override::InferenceResult` in its key though -- `override` should be there as far as [`any(override.overridden_by_const)`](https://github.com/JuliaDebug/Cthulhu.jl/blob/d01b20b6d090867b33525ecf316dfde46031c79f/src/interpreter.jl#L80), and in recent Julia versions that should always be true when we do constant prop': https://github.com/JuliaLang/julia/blob/ac1d69302ce8e682bef3108296f5a6c182b62e5f/base/compiler/abstractinterpretation.jl#L811

**Comment by timholy** [comment]
Thanks! Getting a reproducer is not easy because Simon & I are in the middle of some changes, but a possible reproducer is something like this:

```
]dev Makie
]dev ~/.julia/dev/Makie/CairoMakie
using SnoopCompileCore
using CairoMakie
tinf = @snoopi_deep save("/tmp/test.png", scatter(1:4));
using SnoopCompile
itrigs = inference_triggers(tinf);
sort!(itrigs; by=inclusive);
```
and then peel items off the end of `itrigs` until you hit
```
julia> itrig = itrigs[end-2]
Inference triggered to call Base.Broadcast.materialize(::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1}, Nothing, typeof(CairoMakie.project_rect), Tuple{Base.RefValue{Scene}, Base.RefValue{Symbol}, Vector{GeometryBasics.HyperRectangle{2, Int64}}, Base.RefValue{StaticArrays.SMatrix{4, 4, Float32, 16}}}}) from draw_poly (/home/tim/.julia/dev/Makie/CairoMakie/src/overrides.jl:67) with specialization CairoMakie.draw_poly(::Scene, ::CairoMakie.CairoScreen{Cairo.CairoSurfaceBase{UInt32}}, ::Combined{Makie.poly, Tuple{Vector{Vector{Point{2, Float32}}}}}, ::Vector{GeometryBasics.HyperRectangle{2, Int64}})
```
(it may not be `end-2` for you). Then do
```
julia> ascend(itrig)
Choose a call for analysis (q to quit):
     materialize(::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1}, Nothing, typeof(CairoMakie.project_rect), Tuple{Base.RefValue{Scene}, Base.RefValue{Symbol}, Vector{GeometryBasics.Hyper
 >     draw_poly(::Scene, ::CairoMakie.CairoScreen{Cairo.CairoSurfaceBase{UInt32}}, ::Combined{Makie.poly, Tuple{Vector{Vector{Point{2, Float32}}}}}, ::Vector{GeometryBasics.HyperRectangle{2, Int64}}) at
         draw_poly(::Scene, ::CairoMakie.CairoScreen{Cairo.CairoSurfaceBase{UInt32}}, ::Combined{Makie.poly, Tuple{Vector{Vector{Point{2, Float32}}}}}, ::GeometryBasics.HyperRectangle{2, Int64}) at /home
           draw_plot(::Scene, ::CairoMakie.CairoScreen{Cairo.CairoSurfaceBase{UInt32}}, ::Combined{Makie.poly, Tuple{Vector{Vector{Point{2, Float32}}}}}) at /home/tim/.julia/dev/Makie/CairoMakie/src/over
...
```
select that line, and then select the first call:
```
Select a call to descend into or ↩ to ascend. [q]uit. [b]ookmark.
Toggles: [o]ptimize, [w]arn, [h]ide type-stable statements, [d]ebuginfo, [r]emarks, [e]ffects, [i]nlining costs, [t]ype annotations, [s]yntax highlight for Source/LLVM/Native.
Show: [S]ource code, [A]ST, [T]yped code, [L]LVM IR, [N]ative code
Actions: [E]dit source code, [R]evise and redisplay
Advanced: dump [P]arams cache.
 • %3  = < constprop > getproperty(::Combined{Makie.poly, Tuple{Vector{Vector{Point{2, Float32}}}}},::Core.Const(:model))::Any
   %5  = < constprop > get(::Combined{Makie.poly, Tuple{Vector{Vector{Point{2, Float32}}}}},::Core.Const(:space),::Core.Const(:data))::Any
   %6  = to_value(::Any)::Any
...
```

**Comment by timholy** [comment]
Oh, and this is on Julia commit 258ddc07d413e4c5889a6594a5b93ac6e1508018 (March 12th)

**Comment by aviatesk** [comment]
Hmm, I hit this error:
```julia
julia> tinf = @snoopi_deep save("/tmp/test.png", scatter(1:4));
ERROR: Space screen not recognized. Must be one of (:data, :pixel, :relative, :clip)
```

Any thought?

---

But I'm just okay with this change -- if you don't have time, please go ahead and merge this even without a test.

**Comment by timholy** [comment]
Not sure. If you want to take a look, I put the Project and Manifest file [here](https://gist.github.com/timholy/0461cc9179e54167a337dabf59d9559d).

**Comment by timholy** [comment]
Also, when debugging this I inserted some `@show`s and then discovered the need for this change to `Base`:

```diff
diff --git a/base/show.jl b/base/show.jl
index 8359690034..39802b09a5 100644
--- a/base/show.jl
+++ b/base/show.jl
@@ -2532,7 +2532,7 @@ function show(io::IO, src::CodeInfo; debuginfo::Symbol=:source)
 end
 
 function show(io::IO, inferred::Core.Compiler.InferenceResult)
-    tt = inferred.linfo.specTypes.parameters[2:end]
+    tt = unwrap_unionall(inferred.linfo.specTypes).parameters[2:end]
     tts = join(["::$(t)" for t in tt], ", ")
     rettype = inferred.result
     if isa(rettype, Core.Compiler.InferenceState)
```
Again missing an easy test case though (though I suspect this one would be easier and can give it a try if you want).

**Comment by aviatesk** [comment]
> Not sure. If you want to take a look, I put the Project and Manifest file [here](https://gist.github.com/timholy/0461cc9179e54167a337dabf59d9559d).

These files helped a lot. But I couldn't reproduce the error as I couldn't find the callsite for `Inference triggered to call Base.Broadcast.materialize(::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1}, Nothing, typeof(CairoMakie.project_rect), Tuple{Base.RefValue{Scene}, Base.RefValue{Symbol}, Vector{GeometryBasics.HyperRectangle{2, Int64}}, Base.RefValue{StaticArrays.SMatrix{4, 4, Float32, 16}}}}) from draw_poly (/home/tim/.julia/dev/Makie/CairoMakie/src/overrides.jl:67) with specialization CairoMakie.draw_poly(::Scene, ::CairoMakie.CairoScreen{Cairo.CairoSurfaceBase{UInt32}}, ::Combined{Makie.poly, Tuple{Vector{Vector{Point{2, Float32}}}}}, ::Vector{GeometryBasics.HyperRectangle{2, Int64}})` somehow.

But anyway. Let's move this forward. I may insert a debug statement in the additional fallback code path and see if I hit this problem.

**Pull request merged**